WIP | Dev: Initial bit integration

mtreacy002 opened the pull request:
### Description
Add initial integration for BIT backend to MS backend 

Fixes #1028 

### Type of Change:
- Code
- Quality Assurance

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
1. Complete the local development setup as instructed on [BIT wiki page](https://github.com/anitab-org/bridge-in-tech-backend/blob/develop/.github/ENV_SETUP_INSTRUCTION.md). For ms-backend-server bit version, run it from this PR branch.
2. run BIT backend server FIRST to create the schemas inside postgresql. then run MS backend server from this PR branch.
3. Register a user and see success message.
<img width="1342" alt="Screen Shot 2021-03-03 at 5 39 20 pm" src="https://user-images.githubusercontent.com/29667122/109764288-79a04180-7c47-11eb-8e44-c77e73eb39e3.png">
4. run tests by executing the command `python -m unittest discover tests` and see all test run successfully
<img width="743" alt="Screen Shot 2021-03-03 at 5 49 24 pm" src="https://user-images.githubusercontent.com/29667122/109765500-10213280-7c49-11eb-8e4a-23a90e92d691.png">

5. Login on BIT Swagger UI and make call requests related to MS users service (on gif below GET user/personal_details and GET users/{user_id} were made as examples)
![ezgif com-gif-maker (36)](https://user-images.githubusercontent.com/29667122/111092828-4c2e8e80-858b-11eb-83e5-8b27db9ca1fb.gif)


### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged
- [ ] Update requirements.txt

**Code/Quality Assurance Only**

- [ ] My changes generate no new warnings 
- [ ] New and existing unit tests pass locally with my changes

**Additional note**
For the giithub action to run test successfully, the `main.yml` need to be modified to include adding postgresql plugin and creating test database with the 2 schemas as it was done on Travis.yml on BIT backend repository.
<img width="737" alt="Screen Shot 2021-03-03 at 5 55 38 pm" src="https://user-images.githubusercontent.com/29667122/109766130-def53200-7c49-11eb-9c10-69cd23055080.png">

mtreacy002 committed:
MS Backend server for BIT development purpose

replace db uri

add Flask-Cors to requirements.txt

linting with black formatter

remove fetch env variables duplication

change pull-push branch for github action to bit branch

add postgresql test db

add shell_context_processor

Add -h postgres to psql command

Place PGPASSWORD at the start of psql command

Add env for running test

* Update main.yml (Patch #21)

Co-authored-by: Aditya Kurkure

use non-macOS syntax for db_uri and add db envs

Add postgres db url for github action

Add test for db_test_uri

remove hard coded test db uri

Add psycopg2

Fix db type to postgres+psycopg2

add env to Generate coverage report

move env to global runner

focus pytest cov on tests folder

Remove print command

epicadk commented on app/database/models/mentorship_relation.py:
```diff
@@ -16,10 +16,10 @@ class MentorshipRelationModel(db.Model):
         mentee: relationship between UserModel and mentorship_relation.
-        creation_date: float that defines the date of creation of the mentorship.
-        accept_date: float that indicates the date of acceptance of mentorship.
-        start_date: float that indicates the starting date of mentorship.
```
I honestly don't understand why we use floating point numbers for dates on the mentorship system backend to being with. Should we just add this to the develop branch as well?

mtreacy002 replied:
I understand your point, but this is for @isabelcosta to decide 😉.

epicadk commented on config.py:
```diff
@@ -42,6 +42,7 @@ class BaseConfig(object):
     DB_PASSWORD = os.getenv("DB_PASSWORD")
     DB_ENDPOINT = os.getenv("DB_ENDPOINT")
     DB_NAME = os.getenv("DB_NAME")
```
We have fetched the environment variable here .... (continued below)

mtreacy002 replied:
@epicadk , the one below is for different db_uri (test db_uri). This one is for the main db_uri). The thing is, I'm aiming for better code readability then trying to put them together but makes it less readable 😁.

mtreacy002 replied:
I'll try to refine the code now 😉

epicadk commented on config.py:
```diff
@@ -78,21 +79,23 @@ class BaseConfig(object):
+        db_user_arg=os.getenv("DB_USERNAME"),
+        db_password_arg=os.getenv("DB_PASSWORD"),
+        db_endpoint_arg=os.getenv("DB_ENDPOINT"),
+        db_name_arg=os.getenv("DB_NAME"),
```
Then why not use the DB_NAME variable instead of refetching them from the environment variables?

mtreacy002 replied:
I'm not sure I'm following, 🤔.

epicadk replied:
> I'm not sure I'm following, 🤔.

This was a continuation from the above comment. On line 44 there is a variable called DB_NAME that has this value, then why not use that here instead of fetching the environment variable again.

mtreacy002 replied:
Fixing it now. Thanks for pointing that out 😊

mtreacy002 replied:
@epicadk , done. Can you please check again? Thanks

epicadk replied:
Looks good 💯

epicadk commented on run.py:
```diff
@@ -16,6 +19,8 @@ def create_app(config_filename: str) -> Flask:
 
     migrate = Migrate(app, db)
 
+    cors.init_app(app, resources={r"*": {"origins": "http:localhost:5000"}})
```
Does BIT required access to all endpoints of the mentorship backend or can we give access to specific endpoints instead of all the endpoints?

mtreacy002 replied:
Yes, @epicadk. With users functionaliities and mentorship relation are managed by Mentorship System, BIT does need to have access to all Mentoship System api endpoints 😉.

epicadk replied:
Do you think we should fetch the origins property based on the build type i.e. production, dev, local etc.?

mtreacy002 replied:
@epicadk, I think it's best to keep it as it for now, coz majority of contributors work on local machine atm. When we have remote servers for dev and production, we can revisit this origin property.

mtreacy002 commented:
@epicadk and @annabauza . I'm trying to set up postgresql inside github action but having trouble getting it right. 
<img width="629" alt="Screen Shot 2021-03-04 at 12 42 22 am" src="https://user-images.githubusercontent.com/29667122/109814563-c35a4d80-7c82-11eb-9b61-db4d5475afa3.png">

Here's the modification to [`main.yml`](https://github.com/mtreacy002/mentorship-backend/blob/initial-bit-integration/.github/workflows/main.yml) i've done.
From the PR description above (`Additional Content`) you can see how I've done it using Travis CI
PS: I've never written CI/CD to github action before so, any help would be appreciated.

epicadk commented:
It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?

mtreacy002 commented:
> It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?

can you explain what you mean by setting up on the python code, @epicadk ?

epicadk commented:
> > It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?
> 
> can you explain what you mean by setting up on the python code, @epicadk ?

Like we do for the mentorship system `db.create_all()` over [here](https://github.com/anitab-org/mentorship-backend/blob/479f1c75a8a84a598612ef53f7bb7bc372e13f3e/run.py#L45)

mtreacy002 commented:
> > > It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?
> > 
> > 
> > can you explain what you mean by setting up on the python code, @epicadk ?
> 
> Like we do for the mentorship system `db.create_all()` over [here](https://github.com/anitab-org/mentorship-backend/blob/479f1c75a8a84a598612ef53f7bb7bc372e13f3e/run.py#L45)

I've done that, @epicadk , but still facing error 😓

mtreacy002 commented:
Update @epicadk :
I'm now able to create test db and connect to Postgres in the build 🎉 💃 
But now I'm stuck with `install dependencies` as shown below. Any idea why this happens? Does this have something to do with permission level to the Mentorship System GitHub repo? cc @isabelcosta 
<img width="553" alt="Screen Shot 2021-03-05 at 12 02 34 am" src="https://user-images.githubusercontent.com/29667122/109968908-64124100-7d47-11eb-84ab-6a6011db95b3.png">

I also noticed that when the PR #654 for initial MS migration to GitHub action was merged, the `Run test` workflow did not get triggered. It was only triggered on the PR after  (#655). I wonder why 🤔...
<img width="488" alt="Screen Shot 2021-03-05 at 12 24 02 am" src="https://user-images.githubusercontent.com/29667122/109970694-755c4d00-7d49-11eb-8061-fd0338f66691.png">

annabauza commented:
Maya github have environment variable you can setup, but I don't think you need it here, you should define them on heroku itself

mtreacy002 commented:
> Maya github have environment variable you can setup, but I don't think you need it here, you should define them on heroku itself

@annabauza , are you talking about this error message?  
<img width="553" alt="Screen Shot 2021-03-05 at 12 02 34 am" src="https://user-images.githubusercontent.com/29667122/110036090-98f6b600-7d90-11eb-9182-994c24f0a093.png">

annabauza commented:
Ahh sorry I forgot you have testing database... Ignore me please.

NenadPantelic commented:
@mtreacy002 Try with this:
`pip3 install --no-cache-dir -r requirements.txt` instead of the current pip installation command. I think it is a user permission error.

epicadk commented:
> @mtreacy002 Try with this:
> `pip3 install --no-cache-dir -r requirements.txt` instead of the current pip installation command. I think it is a user permission error.

Yeah that error was fixed. If you check @mtreacy002 's fork I have sent a pr however now the tests are failing. I think the error may be in the database uri that we are providing.

mtreacy002 commented:
Hi @epicadk , I want to show that you've co-authored this pr. How to do this? Should I change manually by replacing name and email address below? If so, can you give your email address here?
<img width="355" alt="Screen Shot 2021-03-06 at 10 18 38 am" src="https://user-images.githubusercontent.com/29667122/110185200-d8db9d00-7e65-11eb-9cc9-675cc17b8016.png">

mtreacy002 commented:
@epicadk , the tests and lint are now passing 🎉 💃 
But the codecov still fails 😓
How do we fix this? Can we remove the codecov from `main.yml` since the repo will do this separately as this is how it is done on `develop` branch.

mtreacy002 commented:
All tests passed now 🎉 💃 
@epicadk can you please review again? Thanks

epicadk commented:
> Hi @epicadk , I want to show that you've co-authored this pr. How to do this? Should I change manually by replacing name and email address below? If so, can you give your email address here?
> <img alt="Screen Shot 2021-03-06 at 10 18 38 am" width="355" src="https://user-images.githubusercontent.com/29667122/110185200-d8db9d00-7e65-11eb-9cc9-675cc17b8016.png">

Haha that's okay once it'll probably get lost in the squashed commits. Glad I could help. 😁

epicadk commented on config.py:
```diff
@@ -132,8 +138,16 @@ class TestingConfig(BaseConfig):
+
+    # uncomment when pushing to upstream for Github action to Run test
+    SQLALCHEMY_DATABASE_URI = (
+        "postgresql://postgres:postgres@localhost:5432/bit_schema_test"
```
We are returning a hard coded string. Maybe if we return a string based using the environment variables, contributors won't have to uncomment this. Do you think that is a good idea?

mtreacy002 replied:
Yep, fix it now by using the existing db_test_uri for non-MacOS lines 😉

epicadk reviewed (approved):
LGTM 👍🏼 . But I have a question about the testing. Will I need to setup BIT on my local machine as well or can I just run the tests on the mentorship system backend?

mtreacy002 commented:
> LGTM 👍🏼 . But I have a question about the testing. Will I need to setup BIT on my local machine as well or can I just run the tests on the mentorship system backend?

If this PR is merged, we can use this branch as the MS backend server for BIT project, instead of setting up a different MS backend from my fork repo. But as BIT project has 3 servers, then yes, you would still need to setup BIT backend so you can test the BIT features. Hope this answer your question 😸 
PS: The condition of BIT backend server must be run first before running the MS backend server on initial setup of BIT project still applies, so the Postgres db gets created properly with the 2 schemas from BIT backend code.

isabelcosta commented on app/api/resources/user.py:
```diff
@@ -337,7 +337,7 @@ def post(cls):
         if result[1] is HTTPStatus.CREATED:
             send_email_verification_message(data["name"], data["email"])
-
+        print(result[0])
```
😝 can you remove this print() statement?

isabelcosta replied:
```suggestion
```
Or is it needed @mtreacy002 ?

isabelcosta replied:
if it's not needed, just remove it and I am happy to merge it after @mtreacy002

mtreacy002 replied:
Ooppss, forgot to take it out 😂. Will remove it asap, @isabelcosta 😉

mtreacy002 replied:
Done, @isabelcosta , please review again when you have time 😉

devkapilbansal commented on .github/workflows/main.yml:
```diff
@@ -5,33 +5,66 @@ name: Run tests
       run: |
-        python -m pip install --upgrade pip
+        sudo -H python -m pip install --upgrade pip
+        pip --version
```
This line is not useful

devkapilbansal replied:
For line no. 56 only

epicadk replied:
> For line no. 56 only

Oops sorry my bad I added that when it wasn't working just to make sure pip is working.

mtreacy002 replied:
Done. I've removed it in the latest push 😉

devkapilbansal commented on app/api/dao/user.py:
```diff
@@ -259,11 +259,11 @@ def update_user_profile(user_id: int, data: Dict[str, str]):
-        if "organization" in data:
-            if data["organization"]:
-                user.organization = data["organization"]
+        if "current_organization" in data:
```
Why this field is renamed? Is it for making it compatible with `BIT`

mtreacy002 replied:
that's right, @devkapilbansal . because bit schema has an attribute called organization as well. So, I renamed it here on MS coz it clashes with the BIT attribute otherwise.

devkapilbansal commented on app/api/models/user.py:
```diff
@@ -47,7 +47,9 @@ def add_models_to_namespace(api_namespace):
         "occupation": fields.String(required=True, description="User occupation"),
-        "organization": fields.String(required=True, description="User organization"),
+        "current_organization": fields.String(
+            required=True, description="User current_organization"
```
```suggestion
            required=True, description="User current organization"
```

devkapilbansal replied:
It would be better if same convention is followed in `models/admin.py` and `models/user.py`.

devkapilbansal replied:
Same for other instances as well

mtreacy002 replied:
> It would be better if same convention is followed in `models/admin.py` and `models/user.py`.

models/admin, @devkapilbansal ? I can't see that file exist in MS backend base code?

> Same for other instances as well

I'm not sure I'm following, can you please elaborate? Which convention you are talking here?

mtreacy002 replied:
done

devkapilbansal commented on config.py:
```diff
@@ -84,15 +85,17 @@ def build_db_uri(
         db_endpoint_arg=DB_ENDPOINT,
         db_name_arg=DB_NAME,
```
If we are not using these variables, then why they are used in function definition?

mtreacy002 replied:
this variables is being used for the linux/windows os, by uncommenting line 131 below and comment out line 129

mtreacy002 replied:
this variables is being used for the linux/windows os, by uncommenting line 131 below and comment out line 129

devkapilbansal commented on tests/test_app_config.py:
```diff
@@ -24,7 +24,6 @@ def test_app_testing_config(self):
         self.assertFalse(application.config["DEBUG"])
         self.assertTrue(application.config["TESTING"])
         self.assertFalse(application.config["SQLALCHEMY_TRACK_MODIFICATIONS"])
-        self.assertEqual("sqlite://", application.config["SQLALCHEMY_DATABASE_URI"])
```
Can we add similar line for `postgres`?

epicadk replied:
I think it would depend on whether the developer is using `postgres` or `postgres+psycopg2`.

mtreacy002 replied:
@devkapilbansal , as per @epicadk said, this is very machine specific. To place it here in the test config means we have to provide separate syntax for each machines. Plus, we will ask contributors to always comment in/out the line/s that are not relevant to their machine when they run the local testing. After that, when they push the PR, they have to always only comment in the line that are going to be used when running test on CI (which is using ubuntu-latest) and comment out the rest, so, I don't think it's ideal to have the db_uri matching syntax to be added to the test.
Unfortunately, atm this is how we do it in the `config.py` for Local and Test config for different OS. Perhaps you can suggest a logic to tackle this issue?

devkapilbansal reviewed (changes requested):
@mtreacy002 although the code looks good, I have some questions/suggestions as follows:-

mtreacy002 commented:
@devkapilbansal . Can you please review again?

mtreacy002 commented:
@epicadk, can you please tell me how to fix the failed codecov? Also, why is the codecov report is not showing for this PR just like it is shown on the other PR for `develop` branch (e.g. #1039)?

epicadk commented:
@mtreacy002 could you avoid force pushing the changes? Kinda makes it harder to review. I had discussed this with @isabelcosta in the QA open session and I think the conclusion was that only while merging the maintainer needs to follow the commit guidelines given in the contributing guidelines. cc @isabelcosta . 😅

mtreacy002 commented:
> @mtreacy002 could you avoid force pushing the changes? Kinda makes it harder to review. I had discussed this with @isabelcosta in the QA open session and I think the conclusion was that only while merging the maintainer needs to follow the commit guidelines given in the contributing guidelines. cc @isabelcosta . 😅

Ok, it makes sense actually 😁. It's just a habit from squashing commits, I guess (which I actually forgot to do on the latest push 🤣). Will just do normal push the next one 😉

mtreacy002 commented:
Oh, on the codecov report, I check the [anitab codecov.io](https://app.codecov.io/gh/anitab-org/mentorship-backend/pulls?page=1&state=open&order=-pullid), and it looks like it's passing, so I'm not sure why it says failed here 🤔
<img width="804" alt="Screen Shot 2021-03-14 at 12 38 22 am" src="https://user-images.githubusercontent.com/29667122/111032012-78002600-845e-11eb-9642-a009cf703951.png">

epicadk commented:
I don't have much knowledge of the subject 😅. I think @devkapilbansal can help out here.

mtreacy002 commented:
> I don't have much knowledge of the subject 😅. I think @devkapilbansal can help out here.

no worries..., @epicadk 😉. @devkapilbansal can you please share some input 🙏.

devkapilbansal reviewed (commented):
@mtreacy002 this PR consists of 5 commits but only one is relevant. All four are possibly because `bit` branch is not updated with `develop` branch. 
So they should be removed.
Otherwise it is good for me.

mtreacy002 commented:
> @mtreacy002 this PR consists of 5 commits but only one is relevant. All four are possibly because `bit` branch is not updated with `develop` branch.
> So they should be removed.
> Otherwise it is good for me.

Done, @devkapilbansal , I removed the other 4 commits. It was initially to update `bit` branch with develop. But the maintainers can do that from their side too 😉

mtreacy002 commented:
@epicadk , should I also integrate your push for docker change in here? Or that can be done after?

epicadk commented:
> @epicadk , should I also integrate your push for docker change in here? Or that can be done after?

It can be done after. I'll send open the pr on this repository.

mtreacy002 commented:
@devkapilbansal and @epicadk , can you please re-approve this PR now so it can be merged? That way the contributors can push directly to upstream `bit` branch instead of to my fork ms repo. Thanks again 😉

epicadk reviewed (approved):
💯

isabelcosta reviewed (approved):
Looks good to me! @mtreacy002 🤗

isabelcosta commented:
@mtreacy002 this is working for your specific case right? Should we test this again? or pass that, because you were already using this for BIT?

devkapilbansal reviewed (approved):
LGTM :+1: 

P.S.:- I haven't tested it

devkapilbansal commented:
> @epicadk, can you please tell me how to fix the failed codecov? Also, why is the codecov report is not showing for this PR just like it is shown on the other PR for `develop` branch (e.g. #1039)?

@mtreacy002 @epicadk looking into the codecov reports, it fails because of changes in `run.py` file. Also, in workflows, branches mentioned is `develop`. That's why codecov report is not shown here

mtreacy002 commented:
> @mtreacy002 this is working for your specific case right? Should we test this again? or pass that, because you were already using this for BIT?

@isabelcosta , I believe since the GitHub action shows that the build is passing here we won't need to do the manual testing again. I also have been using this branch for my BIT for MS backend server and able to use existing features as normal.
PS: I've added a gif on successful calls to MS users REST API from BIT Swagger UI using this PR branch under `How has this been tested` section.


> > @epicadk, can you please tell me how to fix the failed codecov? Also, why is the codecov report is not showing for this PR just like it is shown on the other PR for `develop` branch (e.g. #1039)?
> 
> @mtreacy002 @epicadk looking into the codecov reports, it fails because of changes in `run.py` file. Also, in workflows, branches mentioned is `develop`. That's why codecov report is not shown here

@devkapilbansal, can you tell me where the `develop` branches being mentioned in workflows that causing the codecov fails? as I've changed this to `bit` in the `main.yml` on this pr.
<img width="394" alt="Screen Shot 2021-03-15 at 12 17 18 pm" src="https://user-images.githubusercontent.com/29667122/111091936-a24e0280-8588-11eb-9f66-ab04cfaa4c8a.png">

Are you talking on the codecov.io settings of AnitaB? Perhaps this must be done from the admins side since they have the authority on that?

mtreacy002 commented:
@devkapilbansal , @epicadk , can you please let me know what else I need to do for this PR? If nothing else needed, we can label this issue `Ready to merge` so the maintainer can merge it to the MS `bit` branch.

epicadk commented:
> @devkapilbansal , @epicadk , can you please let me know what else I need to do for this PR? If nothing else needed, we can label this issue `Ready to merge` so the maintainer can merge it to the MS `bit` branch.

We have code approval . Now it needs testing . Busy with college this week i'll test it tonight.

epicadk commented:
Okay I just realized I'll need to have postgreSQL installed to test this which I haven't installed yet.

mtreacy002 commented:
Will close this issue as we're moving to full BIT-MS integration as discussed on BIT Open Session Sunday, 11th April.
